haas/rest.py

```diff
@@ -115,37 +110,113 @@ def foo(bar, baz, quux):
-    `path` is a url path, as recognized by werkzeug's router.
+def _make_schema_for(func):
+    """Build a default schema for the api function `func`.
 
-    `_make_schema_for` will build a schema to validate the body of a request,
-    which will expect the body to be a json object whose keys are (exactly)
-    the set of positional arguments to `func` which cannot be drawn from
-    `path`, and whose values are strings.
+    `_make_schema_for` will build a schema to validate the arguments to an
+    API call, which will expect each argument to be a string. Any arguments
+    not found in the url will be pulled from a JSON object in the body of a
+    request.
 
     If all of the arguments to `func` are accounted for by `path`,
     `_make_schema_for` will return `None` instead.
     """
-    path_vars = [var for (converter, args, var) in parse_rule(path)
-                 if converter is not None]
     argnames, _, _, _ = inspect.getargspec(func)
-    schema = dict((name, basestring) for name in argnames)
-    for var in path_vars:
-        del schema[var]
-    if schema == {}:
-        return None
-    return Schema(schema)
+    return Schema(dict((name, basestring) for name in argnames))
+
+
+def _do_validation(schema, kwargs):
+    """Validate the current request against `schema`.
+
+    `schema` should be a schema as passed to `rest_call`.
+
+    `kwargs` should be the arguments to the API call pulled from the URL.
+
+    If the schema validates, this will return a dictioiniary of *all* of the
+    arguments to the function, both from the URL and the body. The argument
+    `kwargs` may be distructively updated.
+
+    If the schema does not validate, this will raise an instance of
+    `ValidationError`.
+    """
+
+    def _validates(value):
+        """Retrun true if `schema` validates `value`, False otherwise.
+
+        The schema library's `validate` method raises an exception, which is
+        nice for short-circuting when just doing a simple check, but a bit
+        awkward in more complex cases like the below.
+        """
+        try:
+            schema.validate(value)
+        except SchemaError:
+            return False
+        return True
+
+    if schema is None or _validates(kwargs):
+        return kwargs
+
+    # One innocuous reason validation can fail is simply that we
+    # need parameters from the body (in json). Let's get those and
+    # try again.
+    #
+    # Without `force=True`, this will fail unless the
+    # "Content-Type" header is set to "application/json". We'll be
+    # a little more lienient here; we can discuss whether we want
+    # to tighten this up later, but this makes mucking around with
+    # curl less annoying, and is consistent with pre-flask
+    # behavior.
+    body = flask.request.get_json(force=True)
+
+    validation_error = ValidationError(
+        "The request body %r is not valid for "
+        "this request." % body)
+    for k in body.keys():
```

Comment:
It's also arguably clearer this way. I don't *really* have a strong preference, but I'm kind of balking at the level of premature optimization you're talking about.